src/main/java/org/jboss/ejb/client/EJBClientContext.java

```diff
@@ -1148,51 +1155,76 @@ private void waitForClusterTopology(final String clusterName) {
 
     private void attemptReconnections() {
         final CountDownLatch reconnectTasksCompletionNotifierLatch;
-        synchronized(this) {
+        final List<ReconnectHandler> reconnectHandlersToAttempt;
+
+        // check for circumstances where we don't want to attempt reconnection
+        synchronized (this) {
```

Comment:
It was intentional, and i'm not sure of there is a simpler way of arranging it.
When a reconnection is required and there are many incoming invocations, all will be redirected through the "attempt reconnections" code path and each will try to initiate a reconnection attempt. I wanted to make one thread (the first one arriving) not only manage the reconnection attempts, but also cause the other threads to block and then control their release after the reconnection completes, in such a way that they do not again try to attempt re-connections (as it has just been completed). Using the monitor lock didn't seem to allow me the possibility of changing the code path of these latter threads, which is why I opted for a separate AtomicBoolean and ReentrantLock/Condition.